FirebaseSessions/Sources/Settings.swift

```diff
@@ -0,0 +1,167 @@
+  }
+
+  var sessionsEnabled: Bool {
+    guard let enabled = settingsDictionary?["sessions_enabled"]?.boolValue else {
+      return true
+    }
+    return enabled
+  }
+
+  var samplingRate: Double {
+    guard let rate = settingsDictionary?["sampling_rate"]?.doubleValue else {
+      return 1
+    }
+    return rate
+  }
+
+  var sessionTimeout: TimeInterval {
+    guard let timeout = settingsDictionary?["session_timeout"]?.doubleValue else {
+      return 30 * 60
+    }
+    return timeout
+  }
+
+  var isCacheExpired: Bool {
+    guard settingsDictionary != nil else {
+      return true
+    }
+    return isCacheKeyExpired
+  }
+
+  private var cacheDurationSeconds: TimeInterval {
+    guard let duration = settingsDictionary?["cache_duration"]?.doubleValue else {
+      return Settings.cacheDurationSecondsDefault
+    }
+    return duration
+  }
+
+  init(fileManager: SettingsFileManager = SettingsFileManager(), appInfo: ApplicationInfoProtocol) {
+    self.fileManager = fileManager
+    isCacheKeyExpired = false
+    self.appInfo = appInfo
+  }
+
+  func loadCache(googleAppID: String, currentTime: Date) {
+    guard let cacheData = fileManager.data(contentsOf: fileManager.settingsCacheContentPath) else {
+      Logger.logDebug("[Sessions:Settings] No settings were cached")
+      return
+    }
+    guard let parsedDictionary = cacheData.dictionaryValue else {
+      removeCache()
+      return
+    }
+    do {
+      objc_sync_enter(lock)
+      defer { objc_sync_exit(lock) }
+      _settingsDictionary = parsedDictionary
+    }
+    guard let cacheKeyData = fileManager.data(contentsOf: fileManager.settingsCacheKeyPath),
+          let cacheKey = cacheKeyData.cacheKeyValue else {
+      Logger.logError("[Sessions:Settings] Could not load settings cache key")
+      removeCache()
+      return
+    }
+    guard cacheKey.googleAppID == googleAppID else {
+      Logger
+        .logDebug("[Sessions:Settings] Invalidating settings cache because Google App ID changed")
+      removeCache()
+      return
+    }
+    if currentTime.timeIntervalSince(cacheKey.createdAt) > cacheDurationSeconds {
+      Logger.logDebug("[Sessions:Settings] Settings TTL expired")
+      do {
+        objc_sync_enter(lock)
+        defer { objc_sync_exit(lock) }
+        isCacheKeyExpired = true
+      }
+    }
+    if appInfo.synthesizedVersion != cacheKey.appVersion {
+      Logger.logDebug("[Sessions:Settings] Settings expired because app version changed")
+      do {
+        objc_sync_enter(lock)
+        defer { objc_sync_exit(lock) }
+        isCacheKeyExpired = true
+      }
+    }
+  }
+
+  private func removeCache() {
+    objc_sync_enter(lock)
+    defer { objc_sync_exit(lock) }
+    fileManager.removeCacheFiles()
+    isCacheKeyExpired = true
+    _settingsDictionary = nil
+  }
+}
+
+extension ApplicationInfoProtocol {
+  var synthesizedVersion: String { return "\(appDisplayVersion) (\(appBuildVersion))" }
+}
+
+private extension Data {
```

Comment:
I believe it is customary to put extensions at the top of the file

Reply:
Done